### DISCUSSION TITLE
Make the action of saving or exporting something non-blocking

### DISCUSSION THREAD

**Discussion by Speykious** [ppy/osu#18557] [Ideas] [answered]
If there's already a discussion, issue or PR about it, let me know (I've searched and haven't seen it).

Currently, in the editor, pressing the `Save` or the `Export package` button is a blocking operation, and it will freeze the editor until it is finished. The same is true in multiple places, for example when exporting a replay or a skin.
It would be better if the action was non-blocking and there was a visual way to look at the progress of the action. I'm thinking of having a notification for this, I find it perfectly fit for progress-lenient actions.
Not to mention that if implemented correctly, we could do multiple of these actions in a row, like saving the beatmap then immediately exporting the package while it's still saving.

**Comment by Speykious** [comment]
I'm willing to give this a shot as a first contribution, unless you think it would change too many parts of the code.

After some examination of the problem, it got me all the way down to Realm, where the bottleneck is this method:
https://github.com/ppy/osu/blob/45451178dfd08174071ac38d382cc5d490842dc0/osu.Game/Stores/RealmFileStore.cs#L64-L72
It would need to support some kind of way to report progress (if possible) on the action of copying. Maybe a Bindable.

**Comment by Theighlin** [comment, marked as answer]
I think this is #18476

**Reply by Speykious** [reply]
Ah, indeed, thanks. It only mentions skin exporting but it also applies to the cases I mentioned. I'll report what I said here in that issue.
